### DISCUSSION TITLE
ContainerControllerResolver is finding controller by can not get the definition

### DISCUSSION THREAD

**Discussion by eerison** [symfony/symfony#54087] [Q&A]
Hello guys.

I am trying to setup manually dependency injection component like here: https://symfony.com/doc/5.x/create_framework/dependency_injection.html

But as I am using some controllers with dependencies on __construct, then I am replacing `HttpKernel\Controller\ControllerResolver` to `HttpKernel\Controller\ContainerControllerResolver` to get the controller from my container.

the code pass until this point https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php#L52

but when it is trying to get the controller. it throws an exception

```
Class "" does not exist
```
<img width="832" alt="Screenshot 2024-02-27 at 16 58 54" src="https://github.com/symfony/symfony/assets/6358755/15fdaf11-1fda-4a04-b793-14714a822f02">

and debugging I saw that `container->definition->myController->class` is null.
<img width="681" alt="Screenshot 2024-02-27 at 17 01 13" src="https://github.com/symfony/symfony/assets/6358755/0cc0f748-fa7f-42a9-a0ce-2f5491d4484c">

am I missing something?

**Comment by eerison** [comment]
<img width="1148" alt="Screenshot 2024-02-27 at 17 06 18" src="https://github.com/symfony/symfony/assets/6358755/aabafea0-dcd6-4618-bff2-2c9d47e191b9">

**Comment by stof** [comment]
is this something happening in a Symfony fullstack setup (using FrameworkBundle to configure core services and the Kernel class to manage the container builder) or a custom setup ?
In the former case, you should not need to configure a controller resolver at all (FrameworkBundle does it already). In the later case, we don't have enough information to reproduce the issue.

**Comment by eerison** [comment]
I saw that when I add `$container->compile()`, it is adding the class

but I am getting a new issue :(

<img width="817" alt="Screenshot 2024-02-27 at 17 36 55" src="https://github.com/symfony/symfony/assets/6358755/a1d3ca73-1ebf-4a53-bf4c-026f91fb39b4">


it is my container.php

```php
<?php

// @see https://symfony.com/doc/5.x/create_framework/dependency_injection.html
use crmi\Application;
use Symfony\Component\DependencyInjection;
use Symfony\Component\DependencyInjection\Reference;
use Symfony\Component\EventDispatcher;
use Symfony\Component\HttpFoundation;
use Symfony\Component\HttpKernel;
use Symfony\Component\Routing;
use Symfony\Component\Routing\Generator\UrlGenerator;

$container = new DependencyInjection\ContainerBuilder();

$container->register('context', Routing\RequestContext::class);
$container->register('matcher', Routing\Matcher\UrlMatcher::class)
    ->setArguments([$routes, new Reference('context')]);
$container->register('request_stack', HttpFoundation\RequestStack::class);
// $container->register('controller_resolver', HttpKernel\Controller\ControllerResolver::class);
$container->register('controller_resolver', HttpKernel\Controller\ContainerControllerResolver::class)
    ->setArguments([$container]);
$container->register('argument_resolver', HttpKernel\Controller\ArgumentResolver::class);

$container->register('listener.router', HttpKernel\EventListener\RouterListener::class)
    ->setArguments([new Reference('matcher'), new Reference('request_stack')]);

$container->register('listener.response', HttpKernel\EventListener\ResponseListener::class)
    ->setArguments(['UTF-8']);

$container->register('dispatcher', EventDispatcher\EventDispatcher::class)
    ->addMethodCall('addSubscriber', [new Reference('listener.router')])
    ->addMethodCall('addSubscriber', [new Reference('listener.response')]);

$container->register('kernel.crmi_symfony', HttpKernel\HttpKernel::class)
    ->setArguments([
        new Reference('dispatcher'),
        new Reference('controller_resolver'),
        new Reference('request_stack'),
        new Reference('argument_resolver'),
    ]);

$container->register('route.url_generator', UrlGenerator::class)
    ->setArguments([$routes, new Reference('context')]);

$container->set(Application::class, $application);

return $container;
```

calling the compile on my front controller

```php
$container = include __DIR__ . '/container.php';

$request = Request::createFromGlobals();

$loader = new PhpFileLoader($container, new FileLocator(__DIR__));
$loader->load('services.php');

$container->compile();
$kernel = $container->get('kernel.crmi_symfony');
```

**Comment by eerison** [comment]
> is this something happening in a Symfony fullstack setup (using FrameworkBundle to configure core services and the Kernel class to manage the container builder) or a custom setup ?

it is a custom setup, I am replacing a legacy framework to symfony.

**Comment by eerison** [comment]
I could make this work get container from interface's name like

```php
$container->register('controller_resolver', HttpKernel\Controller\ContainerControllerResolver::class)
    ->setArguments([new Reference(ContainerInterface::class)]);
```

but I need to keep services public. at the moment I am facing others issues :D.

if I find other way to make this working with private services I post here.

**Reply by stof** [reply]
To make it work for private services, you need to collect those services in a ServiceLocator and inject that service locator in the ContainerControllerResolver instead of injecting the main container.
